Develop

yura-j opened the pull request:
Заменили лайки на оценки, изменилась система рейтинга, он стал дробным и вычисляется как среднее арифметическое.
Система рекомендаций была улучшена, реализован алгоритм SlopeOne.
Сортировка по популярности была актуализирована в связи с изменением рейтинговой системы.
Произведен рефакторинг системы. Добавлен Optional.
Были написаны тесты Postman коллекции add-marks-like и slope-one, файл /src/main/resources/develop_plus_slope_one.postman_collection.json

Все задачи реализованы. Список:
Добавление режиссёров в фильмы
Удаление фильмов и пользователей
Функциональность  «Отзывы»
Функциональность «Лента событий»
Функциональность «Рекомендации» (Женя)
Функциональность «Поиск»
Функциональность «Общие фильмы»
Вывод самых популярных фильмов по жанру и годам
Рефакторинг: лайки → оценки

GalinaGitH committed:
add-score_and_slope

fidgo committed:
refactor: slope-one simple

fidgo committed:
refactor: optional

fidgo committed:
feat:add postman test slope one

yura-j committed:
feat: refactoring

yura-j committed:
feat: Postman collection

yura-j committed:
feat: Postman collection

yura-j committed:
Merge pull request #27 from GalinaGitH/add-marks

Add marks

yura-j committed:
fix:sql lower_case

vlad-iv commented on src/main/java/ru/yandex/practicum/filmorate/controllers/FilmController.java:
```diff
@@ -49,11 +49,12 @@ public Film updateFilm(@Valid @RequestBody Film film) {
     @PutMapping("/films/{filmId}/like/{userId}")
-    public void addLikes(@PathVariable long filmId, @PathVariable long userId) {
+    public void addLikes(@PathVariable long filmId, @PathVariable long userId, @RequestParam(defaultValue = "10",
+            required = false) @Positive Integer score) {
```
Можно добавить Max 10 ограничение

vlad-iv commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
```diff
@@ -32,13 +32,4 @@ public class Film {
     private long duration; // длительность
```
Предлагаю рефакторинг, добавить расчетное поле RATING (AVG (LIKES.SCORE))
так как лайков много, каждый раз расчет будет медленный

vlad-iv commented on src/main/java/ru/yandex/practicum/filmorate/storage/impl/DirectorDbStorage.java:
```diff
@@ -125,19 +127,22 @@ public List<Film> getSortedFilmsByYearOfDirector(int idDirector) {
                 "WHERE FILM_ID IN (SELECT FILM_ID FROM FILM_DIRECTORS WHERE DIRECTOR_ID = ?) " +
                 "ORDER BY FILM_RELEASE_DATE;";
```
сортировка по году  year(FILM_RELEASE_DATE)

vlad-iv commented on src/main/java/ru/yandex/practicum/filmorate/storage/impl/FilmDbStorage.java:
```diff
@@ -46,97 +44,107 @@ public Film create(Film film) {
-    }
-
     @Override
     public List<Film> getLikedByUser(long userId) {
```
не используется, удаляем

vlad-iv commented on src/main/java/ru/yandex/practicum/filmorate/storage/impl/FilmDbStorage.java:
```diff
@@ -46,97 +44,107 @@ public Film create(Film film) {
     }
 
     @Override
     public List<Film> getLikedByUserSortedPopular(long userId) {
```
Заменить на реализацию findCommonFilms на SQL
так же сортировка нужно по RATING

vlad-iv commented on src/main/java/ru/yandex/practicum/filmorate/storage/impl/FilmDbStorage.java:
```diff
@@ -46,97 +44,107 @@ public Film create(Film film) {
+    }
+
+    @Override
+    public List<Film> getFilmsFromIds(List<Long> idFilms) {
```
общепринятое название такого метода
findById

vlad-iv commented on src/main/java/ru/yandex/practicum/filmorate/storage/impl/FilmDbStorage.java:
```diff
@@ -46,97 +44,107 @@ public Film create(Film film) {
+                "WHERE FILM_ID IN (%s)", sql);
+        int[] argTypes = new int[idFilms.size()];
+        Arrays.fill(argTypes, Types.BIGINT);
+        List<Film> films = jdbcTemplate.query(sql, idFilms.toArray(), argTypes, this::mapRowToFilm);
```
так не будет работать? 
dbcTemplate.query(sql, this::mapRowToFilm, idFilms.toArray());
кажется argTypes  лишнее

vlad-iv commented on src/main/java/ru/yandex/practicum/filmorate/storage/impl/FilmDbStorage.java:
```diff
@@ -152,18 +160,19 @@ public List<Film> search(String query, List<String> by) {
                 .map(Map.Entry::getValue)
```
Хорошая практика использовать enum для by
например Set<FilmSearchBy> by

vlad-iv commented on src/main/java/ru/yandex/practicum/filmorate/storage/impl/LikesDbStorage.java:
```diff
@@ -26,37 +26,57 @@ public LikesDbStorage(JdbcTemplate jdbcTemplate) {
 
     @Override
     public List<Like> getLikes(long filmId, long userId) {
```
Выглядит как getLike
такой метод точно нужен?
Ожидалась что работа с лайками будет в БД на SQL

vlad-iv replied:
тоже лучше уйти от getAllLikes() 
так как лайков очень много и их загрузка очень медленная
заменить SQL

vlad-iv commented on src/main/java/ru/yandex/practicum/filmorate/storage/impl/LikesDbStorage.java:
```diff
@@ -87,7 +107,7 @@ public List<Film> findPopularFilmsByYear(Integer limit, int year) {
                 "WHERE CAST(EXTRACT(YEAR FROM FILMS.FILM_RELEASE_DATE) AS INTEGER) = ? " +
                 "GROUP BY FILMS.FILM_ID " +
-                "ORDER BY COUNT(DISTINCT l.USER_ID) DESC " +
+                "ORDER BY AVG (L.SCORE) DESC " +
```
AVG(SCORE) везде через рефакторинг заменяем на FILM.RATING предварительно рассчитанный на каждый лайк-дизалйк

vlad-iv replied:
в findPopularFilmsByGenre пропущено ORDER BY AVG (L.SCORE) DESC  -> ORDER BY FILMS.RATING DESC

vlad-iv commented on src/main/java/ru/yandex/practicum/filmorate/storage/impl/LikesDbStorage.java:
```diff
@@ -26,37 +26,57 @@ public LikesDbStorage(JdbcTemplate jdbcTemplate) {
     }
 
     @Override
     public void removeLikes(long filmId, long userId) {
```
может  removeLike , ведь filmId и userId - это ключ
пользователь не может задать больше 1 лайка фильму

vlad-iv replied:
добавить вызов приватного метода
updateRating(long filmId)  в  removeLike

vlad-iv commented on src/main/java/ru/yandex/practicum/filmorate/storage/impl/LikesDbStorage.java:
```diff
@@ -26,37 +26,57 @@ public LikesDbStorage(JdbcTemplate jdbcTemplate) {
+    }
+
+    @Override
+    public void updateLikes(long filmId, long userId, Integer score) {
```
updateLike

vlad-iv replied:
добавить вызов приватного метода
updateRating(long filmId)  в updateLike

vlad-iv commented on src/main/java/ru/yandex/practicum/filmorate/storage/impl/LikesDbStorage.java:
```diff
@@ -26,37 +26,57 @@ public LikesDbStorage(JdbcTemplate jdbcTemplate) {
+    }
+
+    @Override
+    public void addLikes(long filmId, long userId, Integer score) {
```
addLike

vlad-iv replied:
добавить вызов приватного метода
updateRating(long filmId)  в addLike

vlad-iv commented on src/main/java/ru/yandex/practicum/filmorate/storage/impl/ReviewDbStorage.java:
```diff
@@ -121,19 +124,20 @@ private Review mapRowToReview(ResultSet resultSet, int rowNum) throws SQLExcepti
      * Если как бесполезный, то уменьшает на 1.
      */
     private int getUsefulRate(long reviewId) {
```
рассчет REVIEW.useful  ожидалось через SQL 
update REVIEWS  set useful = ...
на каждый лайк или его удаление

vlad-iv commented on src/main/java/ru/yandex/practicum/filmorate/storage/impl/ReviewDbStorage.java:
```diff
@@ -53,16 +54,15 @@ public Review create(Review review) {
-        String sqlQuery = "update REVIEWS set " +
-                "IS_POSITIVE = ?,CONTENT = ? " +
-                "where REVIEW_ID= ?";
+        String sqlQuery = " UPDATE REVIEWS SET " +
```
Общая практика update в storage пришется на все поля
а в сервисе, мы по ид получаем запись из БД , в ней устанавливем только новые значение для IS_POSITIVE ,CONTENT . после этого обновляем

yura-j committed:
refactor:
+1. Можно добавить Max 10 ограничение
+3. Cортировка по году year(FILM_RELEASE_DATE)
+4. Не используется, удаляем return jdbcTemplate.query(sqlQuery, this::mapRowToFilm, id, id, id)
+6. Общепринятое название такого метода
findById
+9. Выглядит как getLike
такой метод точно нужен?
Ожидалась что работа с лайками будет в БД на SQL
тоже лучше уйти от getAllLikes()
так как лайков очень много и их загрузка очень медленная
заменить SQL

+11. В findPopularFilmsByGenre пропущено ORDER BY AVG (L.SCORE) DESC -> ORDER BY FILMS.RATING DESC
+12. Может removeLike , ведь filmId и userId - это ключ
пользователь не может задать больше 1 лайка фильму
добавить вызов приватного метода
updateRating(long filmId) в removeLike
+13. updateLike
добавить вызов приватного метода
updateRating(long filmId) в updateLike
+14. addLike
добавить вызов приватного метода
updateRating(long filmId) в addLike

fidgo committed:
refactor: film rating

fidgo committed:
fix: findCommonFilms

fidgo committed:
refactor: remove LIKES from SqlQueries, remove dont used method

fidgo committed:
refactor: review.useful add

yura-j committed:
style: delete unused, reformat, imports.

yura-j commented:
Все что можно пофиксили, спасииибо за ревью))

yura-j committed:
fix: reviewUpdate method fixed

yura-j commented:
review update fixed
